File: 3/pagerank.java
import java.util.ArrayList;
import java.util.Scanner;
import java.io.*;
import java.text.DecimalFormat;

public class pagerank {
    static ArrayList<link> linkek = new ArrayList<link>();

    public static void main(String[] args){

	double d;

	beolvasas();						
	
	Scanner scan = new Scanner(System.in);
	
	System.out.print("Add meg a 'd' erteket: ");
	d = scan.nextDouble();

	for(int i = 0; i < linkek.size(); i++){			
	    System.out.print(linkek.get(i).nev + " : ");
	    for(int j = 0; j < linkek.get(i).mutat.size(); j++){
		System.out.print(linkek.get(i).mutat.get(j) + ", ");
	    }
	    System.out.println();	
	}
	double[][] link_matrix = new double[linkek.size()][linkek.size()];
	double[] PR = new double[linkek.size()];
	double[] PRs = new double[linkek.size()];

	DecimalFormat ket_tizedes = new DecimalFormat("#.##");	

	for(int i = 0; i < linkek.size(); i++){
	    PR[i] = 0;
	    PRs[i] = 1.0 / linkek.size();
	    for(int j = 0; j < linkek.size(); j++){
		boolean benne_van = false;
		for(int k = 0; k < linkek.get(j).mutat.size(); k++){
		    if(linkek.get(j).mutat.get(k).equals(linkek.get(i).nev)){
			benne_van = true;
			link_matrix[i][j] = 1.0 / linkek.get(j).mutat.size();
			break;
		    }
		}
		if(!benne_van){
		    link_matrix[i][j] = 0.0;
		}
		System.out.print(ket_tizedes.format(link_matrix[i][j]) + "\t");
	    }
	    System.out.println();
	}

	while(true){

	    for(int i = 0; i < linkek.size(); i++){
		double temp = 0;
		for(int j = 0; j < linkek.size(); j++){
		    temp += link_matrix[i][j]*PRs[j];
		}
		PR[i] = (1-d)/linkek.size()+d*temp;
	    }
	    if(tavolsag(PR, PRs) < 0.0000000001){
		break;
	    }
	    for(int i = 0; i < linkek.size(); i++){
		PRs[i] = PR[i];
	    }
	}
	for(int i = 0; i < linkek.size(); i++){
	    System.out.println((i+1) + ".: " + PR[i]);
	}

    }

    static void beolvasas(){
	BufferedReader br = null;
	try {
    	    File file = new File("adatok.txt");
            br = new BufferedReader(new FileReader(file));

    	    String temp;
   	    while ((temp = br.readLine()) != null) {
		linkek.add(new link(temp));
		while((temp = br.readLine()) != null && temp.length() != 0){
		    linkek.get(linkek.size()-1).mutat.add(temp);
		}
            }
	    br.close();
	} catch (IOException e) {
	    System.err.println("Nem található a fájl ...");
	}
    }
    static double tavolsag(double[] PR, double[] PRs){
	double osszeg = 0;

	for(int i = 0; i < linkek.size(); i++){
	    osszeg += Math.abs(PR[i]-PRs[i]);
	}
	return osszeg;
    }
}

File: 3/pagerank.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>

int beolvasas_1(int* count, int* max_mutat, int* max_hossz);
double tavolsag(double* PR, double* PRs, int size);

int main(){
    int count = 0, max_mutat = 0, max_hossz = 0;
    if(!beolvasas_1(&count, &max_mutat, &max_hossz)){
	printf("Nincs meg a fajl...\n");
	return 0;
    }
    double d = 0.98;
    char nev[count][max_hossz];			
    int mutat_size[count];			
    char mutat[count][max_mutat][max_hossz];	
						
    printf("Adja meg a d erteket: ");
    scanf("%lf", &d);

    FILE* f;
    f = fopen("adatok.txt", "r");
    char temp[max_hossz];
    
    for(int i = 0; i < count; i++){

	fgets(temp, max_hossz+1, f);
	temp[strlen(temp)-1] = '\0';
	strcpy(nev[i], temp);

	int j = 0;
	while(fgets(temp, max_hossz+1, f) != NULL && strlen(temp) != 1){
	    temp[strlen(temp)-1] = '\0';
	    strcpy(mutat[i][j], temp);

	    j++;
	}
	mutat_size[i] = j;
    }
    fclose(f);
						
    for(int i = 0; i < count; i++){
	printf("%s: ", nev[i]);
	for(int j = 0; j < mutat_size[i]; j++){
	    printf(" %s", mutat[i][j]);
	}
	printf("\n");
    }
						
    double link[count][count];
    double PR[count];
    double PRs[count];
						
    for(int i = 0; i < count; i++){
	PR[i] = 0.0;
    	PRs[i] = 1.0 / count;
	for(int j = 0; j < count; j++){
	    int benne_van = 0;
	    for(int k = 0; k < mutat_size[j]; k++){
		if(strcmp(nev[i], mutat[j][k]) == 0){
		    benne_van = 1;
		    link[i][j] = 1.0 / mutat_size[j];
		    break;
		}
	    }
	    if(!benne_van){
		link[i][j] = 0;
	    }
	    printf("%.2f \t",  link[i][j]);
	}
	printf("\n");
    }
						//PR számolása
    for(;;){
	for(int i = 0; i < count; i++){
	    double temp = 0.0;
	    for(int j = 0; j < count; j++){
		temp += link[i][j]*PRs[j];
	    }
	    PR[i] = d*temp+(1-d)/count;
	}
	if(tavolsag(PR, PRs, count) < 0.000000001){
	    break;
	}
	for(int i = 0; i < count; i++){
	    PRs[i] = PR[i];
	}
    }
    double sum = 0;
    for(int i = 0; i < count; i++){
  	printf("%d.: %.8f\n", i+1, PR[i]);
	sum += PR[i];
    }
    printf("Osszeg: %.4f\n", sum);

    return 0;
}

int beolvasas_1(int* count, int* max_mutat, int* max_hossz){
    FILE* f;
    f = fopen("adatok.txt", "r");
    if(f == NULL){
	return 0;
    }
    int mutat_count = 0;
    char temp[100];
    
    while(fgets(temp, 100, f)){
	*count += 1;
	mutat_count = 0;
	if(strlen(temp) > *max_hossz){
	    *max_hossz = strlen(temp);	
	}
	while(fgets(temp, 100, f) && strlen(temp) != 1){
	    mutat_count++;
	}
	if(mutat_count > *max_mutat){
	    *max_mutat = mutat_count;
	}
    }
    fclose(f);
    return 1;
}

double tavolsag(double* PR, double* PRs, int size){

    double osszeg = 0;
    for(int i = 0; i < size; i++){
	osszeg += (PR[i]-PRs[i])*(PR[i]-PRs[i]);
    }
    return sqrt(osszeg);
}